Try to improve the documentation of `filter()` and `filter_map()`.

vext01 opened the pull request:
I believe the documentation is currently a little misleading.

For example, in the docs for `filter()`:

> If the closure returns `false`, it will try again, and call the closure on
> the next element, seeing if it passes the test.

This kind of implies that if the closure returns true then we *don't* "try
again" and no further elements are considered. In actuality that's not the
case, every element is tried regardless of what happened with the previous
element.

This change tries to clarify that by removing the uses of "try again"
altogether.

rust-highfive commented:
r? @cramertj

(rust_highfive has picked a reviewer for you, use r? to override)

vext01 commented:
I've just noticed that I wrapped my lines longer than before. Let me know if that's an issue and I'll fix it.

Dylan-DPC commented:
r? @steveklabnik

steveklabnik commented:
> I've just noticed that I wrapped my lines longer than before. Let me know if that's an issue and I'll fix it.

What column are they wrapped to? Standard library, IIRC, uses 80.

vext01 commented:
Re-wrapped to 80 and made the docs for `filter_map()` yet more concise.

What do you think?

vext01 commented:
It occurs to me that it might be better to say `returned iterator` instead of `filtered result`? Thoughts?

vext01 commented:
Having thought a little more, I prefer "returned iterator". Fix pushed.

OK to squash?

jyn514 commented on library/core/src/iter/traits/iterator.rs:
```diff
@@ -727,21 +727,10 @@ pub trait Iterator {
-    /// closure returns [`None`], it will try again, and call the closure on the
-    /// next element, seeing if it will return [`Some`].
-    ///
-    /// Why `filter_map` and not just [`filter`] and [`map`]? The key is in this
-    /// part:
-    ///
-    /// [`filter`]: #method.filter
-    /// [`map`]: #method.map
-    ///
-    /// > If the closure returns [`Some(element)`][`Some`], then that element is returned.
-    ///
-    /// In other words, it removes the [`Option<T>`] layer automatically. If your
-    /// mapping is already returning an [`Option<T>`] and you want to skip over
-    /// [`None`]s, then `filter_map` is much, much nicer to use.
```
Where did this section go? This seems useful.

vext01 replied:
I found this part a bit too verbose/conversational, and so condensed it into the main description to make something more concise:

> If the closure returns `Some(element)`, then the mapped element (without the `Some` wrapper) will be present in the returned iterator.

The two paragraphs I removed are captured by `without the Some wrapper` part.

jyn514 replied:
But now there's no discussion of _why_ you'd use this method, just what it does. Half the point of documentation is to tell you why you'd use this API.

bors commented:
:umbrella: The latest upstream changes (presumably #76265) made this pull request unmergeable. Please resolve the merge conflicts.

vext01 commented:
Rebased to resolve merge conflict. [Another change(https://github.com/rust-lang/rust/pull/76265/files#diff-7db43b364e0fe2bf6f24912d40f5e056L737) updated the links which I had removed anyway.

hbina commented:
Why not something like "The returned iterator will only contain elements that returns true when applied to the predicate". Short and no double negatives with "false" and "not"

jyn514 commented:
You removed some links also that need to be added back:
```
error: unresolved link to `filter`
   --> library/core/src/iter/traits/iterator.rs:743:44
    |
743 |     /// Here's the same example, but with [`filter`] and [`map`]:
    |                                            ^^^^^^^^ unresolved link
    |
    = note: `-D broken-intra-doc-links` implied by `-D warnings`
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

error: unresolved link to `map`
   --> library/core/src/iter/traits/iterator.rs:743:59
    |
743 |     /// Here's the same example, but with [`filter`] and [`map`]:
    |                                                           ^^^^^ unresolved link
    |
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
```
I still consider https://github.com/rust-lang/rust/pull/75949#discussion_r482935537 a blocking concern though.

vext01 commented:
I've had a shot at addressing comments.

There's a curious merge conflict (for a path I've not touched), but we can look at that later.

Are we getting any closer?

jyn514 commented:
> There's a curious merge conflict (for a path I've not touched), but we can look at that later.

You updated submodules, I'm not quite sure how to fix it. Maybe `git reset master src/{doc,edition-guide,reference,rust-by-example} src/tools/{cargo,miri,rust-analyzer}`?

jyn514 commented on library/core/src/iter/traits/iterator.rs:
```diff
@@ -719,24 +717,17 @@ pub trait Iterator {
-    /// returns [`Some(element)`][`Some`], then that element is returned. If the
-    /// closure returns [`None`], it will try again, and call the closure on the
-    /// next element, seeing if it will return [`Some`].
+    /// Given an element the closure must return `Some(value)` or `None`. The
+    /// returned iterator will only yield a value when the closure returns
+    /// `Some`. The values inside the `Some`s are yielded.
```
This seems a little verbose, but I'm not sure how better to word it. It seems like it's saying the same thing 3 times sort of.

vext01 replied:
I spent quite some time trying to get these three sentences down to 2 (or even better one), but failed.

It's hard to explain concisely that it is not the element that is returned, but the innards of the `Some` that is returned.

I'm open to suggestions, but also I'd like to move forward. You can always tweak this in a later PR.

vext01 replied:
Would something like this be any better?

> The returned iterator yields only the `value`s for which the supplied closure returns `Some(value)`.

jyn514 replied:
Ooh, I like that!

jyn514 commented on library/core/src/iter/traits/iterator.rs:
```diff
@@ -719,24 +717,17 @@ pub trait Iterator {
     ///
-    /// Why `filter_map` and not just [`filter`] and [`map`]? The key is in this
-    /// part:
+    /// `filter_map` can be used to make chains of [`filter`] and [`map`] more
+    /// concise. The example below shows how a `map().filter().map()` can be
+    /// shortened to a single call to `filter_map`.
```
This is a lot better, but still leaves most of the discussion to the example instead of the prose. Maybe that's not a bad thing?

vext01 replied:
That was my thinking. The existing example is really good, and extra words just dilutes it.

jyn514 replied:
Sounds good to me!

jyn514 commented:
r=me without the submodule updates. Not sure how exactly to fix those ... maybe something like this?
```sh
git checkout filter-docs
git reset fd001558c8f077da2e9e645c6667881b48524e72
git add library/core/src/iter/traits/iterator.rs
git commit -m "Address comments"
git push -f
```

vext01 commented:
@jyn514 I think we raced. I made a suggestion in the above thread at the same time as it was marked resolved.

vext01 commented:
I think this should be good.

Assuming all is well, can I squash?

rust-log-analyzer commented:
The job `mingw-check` of your PR [failed](https://github.com/rust-lang/rust/runs/1075944755) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang/rust/commit/a40d9c72d2a6e572242aad7bd313ee01d9bc3dfa/checks/1075944755/logs), [raw log](https://github.com/rust-lang/rust/commit/a40d9c72d2a6e572242aad7bd313ee01d9bc3dfa/checks/1075944755/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Initialized empty Git repository in /home/runner/work/rust/rust/.git/
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/75949/merge:refs/remotes/pull/75949/merge
```

</details><p></p>

[I'm a bot](https://github.com/rust-lang/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-lang/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

jyn514 commented:
Looks good to me but you still need to undo the submodule changes: https://github.com/rust-lang/rust/pull/75949#issuecomment-687631020

vext01 commented:
Ugh. I thought I had removed those.

I *think* this is the desired result (with a squash also).

jyn514 commented:
@bors r+ rollup

Thanks for the PR! (and for persisting through the trouble with submodules)

bors commented:
:pushpin: Commit 8af85fa1f73428888730b05e48291c73897dd44d has been approved by `jyn514`

<!-- @bors r=jyn514 8af85fa1f73428888730b05e48291c73897dd44d -->
<!-- homu: {"type":"Approved","sha":"8af85fa1f73428888730b05e48291c73897dd44d","approver":"jyn514"} -->

bors commented:
:hourglass: Testing commit 8af85fa1f73428888730b05e48291c73897dd44d with merge 642ebd19e1bccb1179d6acee8af7e56f8920b242...
<!-- homu: {"type":"BuildStarted","head_sha":"8af85fa1f73428888730b05e48291c73897dd44d","merge_sha":"642ebd19e1bccb1179d6acee8af7e56f8920b242"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36450)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36450","builder_name":"checks-azure"} -->

rust-log-analyzer commented:
Your PR [failed](https://github.com/rust-lang-ci/rust/runs/1077859896) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/642ebd19e1bccb1179d6acee8af7e56f8920b242/checks/1077859896/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/642ebd19e1bccb1179d6acee8af7e56f8920b242/checks/1077859896/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[group]Run exit 1
exit 1
shell: /bin/bash --noprofile --norc -e -o pipefail {0}
##[endgroup]
##[error]Process completed with exit code 1.
```

</details><p></p>

[I'm a bot](https://github.com/rust-lang/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-lang/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

jyn514 commented:
@bors retry

network error:
```
curl: (56) SSLRead() return error -9806
clang+llvm-10.0.0-x86_64-apple-darwin/include/clang/AST/ExprObjC.h: Lzma library error:  No progress is possible
tar: Error exit delayed from previous errors.
```

bors commented:
:hourglass: Testing commit 8af85fa1f73428888730b05e48291c73897dd44d with merge 5d74e880061c250936105125a2ffa901f315aaaa...
<!-- homu: {"type":"BuildStarted","head_sha":"8af85fa1f73428888730b05e48291c73897dd44d","merge_sha":"5d74e880061c250936105125a2ffa901f315aaaa"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1078638124), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36455)
Approved by: jyn514
Pushing 5d74e880061c250936105125a2ffa901f315aaaa to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jyn514","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36455","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1078638124"},"merge_sha":"5d74e880061c250936105125a2ffa901f315aaaa"} -->
